### ISSUE TITLE
Insight into proper error handling

### ISSUE THREAD

**Issue by JacksonMaxfield** [napari/cookiecutter-napari-plugin#4]
Given that plugins are supposed to support a list of strings input, should we error if any single one of those images have differing shape from the others?

**Comment by tlambert03** [comment]
Hi @JacksonMaxfield,
The list-of-strings input was a later addition to the reader plugin spec, to accommodate our own internal needs to allow the user to provide a list of similar files for napari to stack (conversation starts https://github.com/napari/napari/pull/937#issuecomment-595875174).  It was decided there that plugin readers will only be handed a list of strings if the user has passed the `stack=True` argument to `napari.Viewer.add_path` (or `napari.view_path`).  So, that means you _should_ be able to assume that they are stackable (at least, the user has stated so) ... and it would be totally fine for you throw an error if they are not (and indeed, [this line](https://github.com/napari/cookiecutter-napari-plugin/blob/bd776118c9ab1c227d08963cba34219f06ebc1b5/napari-%7B%7Bcookiecutter.plugin_name%7D%7D/napari_%7B%7Bcookiecutter.module_name%7D%7D.py#L50) in the example plugin would throw an error) if any image had a different shape from the others.

That said... "technically" readers can do whatever they want with that list of paths.  So if for some reason you didn't _want_ to error, you aren't obligated.

**Comment by JacksonMaxfield** [comment]
Gotcha -- cool. While I have you, the comment in the cookiecutter about "metadata being optional kwargs for the corresponding viewer.add_* method"

Does that mean I need to specify the method then the kwargs?
```python
meta = {
    "view_image": {
        "channel_axis": 1,
        # etc
    }
}
```

**Comment by tlambert03** [comment]
ah, no... you don't specify the layer type or view method in meta dict.  That's the _third_ (optional) member of each tuple in the list your return (if a third item is not supplied, it is assumed to be an image layer).  so in your example it would be `meta = {'channel_axis': 1}`. 
The following are all valid return values from a reader function (where `data` is some np.array or other appropriate data type for the corresponding layer):

```python
# image layer assumed by default
[(data,)]
[(data, {"channel_axis": 1, "name": "My Image"})]
# non image layer
[(data, {"edge_width": 2, "edge_color": "red"}, "points")]
# combination of layer types
[(data,), (label_data, {}, "labels")]
```

basically, the 1st, 2nd, and 3rd members of each tuple should be appropriate for the corresponding positional arguments to [`Viewer._add_layer_from_data()`](https://github.com/napari/napari/blob/ea7ec36eb8956329e7199680721b8bc4d58aa2e2/napari/components/add_layers_mixin.py#L786-L829)

lastly, this PR hasn't merged yet, but there are more docs that you may not have seen yet [here](https://github.com/napari/napari/blob/0b82b863b186e3164905e5c77aa4cf13acff87bb/docs/source/plugins/for_plugin_developers.rst).

**Comment by JacksonMaxfield** [comment]
Thanks for all the help @tlambert03!

While the full plugin interface and plugin manager haven't been merged yet I was easily able to get setup with two variants of an [`aicsimageio` plugin](https://github.com/AllenCellModeling/aicsimageio/pull/87).

Some minor bugfixes to come that are unrelated to that PR haha.

Some more questions on plugin metadata passing:
Two really strong pieces of metadata for us are `axis_labels` and `ndisplay`, I don't see a way to expose those through the plugin interface unfortunately. Have any ideas?

**Comment by JacksonMaxfield** [comment]
Oh and, because our single module exposes two different plugin variants, I tried shuffling the order of the entry points in the `setup.py` but it didn't seem to matter. Is there some method to the madness that is the "how the preferred napari plugin is chosen"?

**Comment by tlambert03** [comment]
> Two really strong pieces of metadata for us are `axis_labels` and `ndisplay`, I don't see a way to expose those through the plugin interface unfortunately. Have any ideas?

ahhh... Short answer is "not yet".  There was a _lot_ of discussion about this early on (see conversation starting around allowing plugins to set global viewer/dims parameters in addition to layer parameters https://github.com/napari/napari/pull/891#issuecomment-576082264).  We ended up avoiding it for now... but we're always open to discussions.  If that is a key thing for you, I'd definitely recommend opening a new question or feature request at the napari repo (not this one) just to get the conversation (re)started.  you can link to this conversation and perhaps also the stuff at #891.

> Is there some method to the madness that is the "how the preferred napari plugin is chosen"?

This is also a work in progress.  A little background: with [`pluggy`](https://github.com/pytest-dev/pluggy), plugin calls come in two flavors [`firstresult=True`](https://pluggy.readthedocs.io/en/latest/#first-result-only), in which the first plugin to provide a non-null result is the one that gets used, and the default mode in which _all_ plugins are allowed to return a result.  For `napari_get_reader`, we use `firstresult=True`... so, as you observe, the "first" reader wins.  by default, pluggy plugins are called in ["last-in-first-out" order](https://pluggy.readthedocs.io/en/latest/#call-time-order) (where "in" is defined by the order of registration... maybe alphabetical?).  Anyway, I'm working on the ability to sort reader plugin calls after registration here: https://github.com/napari/napari/pull/1023

**Comment by JacksonMaxfield** [comment]
Thank you again for all the info! I think for now it is okay for us to not have them and I understand the reasoning behind not letting plugins affect global viewer state.

**Comment by sofroniewn** [comment]
I have opened a dedicated issue around viewer keyword args from reader plugins as it has come up again, so we should at least have one central place for discussion https://github.com/napari/napari/issues/1077
